src/components/tests/SourcesTree.spec.js

```diff
@@ -85,4 +86,622 @@ describe("SourcesTree", () => {
+      });
+
+      it("recreates tree if debugeeUrl is changed", async () => {
+        props.sources = sourceThree;
+        props.debuggeeUrl = "mozilla";
+
+        expect(currState.uncollapsedTree.contents[0].contents).toHaveLength(3);
+
+        await component.setProps({
+          ...props
+        });
+
+        expect(
+          component.state("uncollapsedTree").contents[0].contents
+        ).toHaveLength(1);
+      });
+    });
+
+    describe("updates list items", () => {
+      beforeEach(() => {
+        props.selectLocation.mockClear();
+      });
+
+      it("updates list items if shownSource changes", async () => {
+        props.shownSource = "http://mdn.com/three.js";
+        await component.setProps({
+          ...props
+        });
+        expect(component).toMatchSnapshot();
+        expect(props.selectLocation).toHaveBeenCalledWith({
+          sourceId: "server1.conn13.child1/41"
+        });
+      });
+    });
+
+    describe("updates highlighted items", () => {
+      it("updates highlightItems if selectedSource changes", async () => {
+        props.selectedSource = sourceThree;
+        await component.setProps({
+          ...props
+        });
+        expect(component).toMatchSnapshot();
+      });
+    });
+  });
+
+  describe("focusItem", () => {
+    it("update the focused item", async () => {
+      const { component } = render({});
+
+      expect(component.state("focusedItem")).toEqual(null);
+
+      await component.instance().focusItem(itemOne);
+
+      expect(component.state("focusedItem")).toEqual(itemOne);
+    });
+  });
+
+  describe("with custom root", () => {
+    const { component, props } = render({
+      projectRoot: "mdn",
+      clearProjectDirectoryRoot: jest.fn()
+    });
+
+    afterEach(() => {
+      props.clearProjectDirectoryRoot.mockClear();
+    });
+
+    it("renders custom root source list", async () => {
+      expect(component).toMatchSnapshot();
+    });
+
+    it("renders empty custom root source list", async () => {
+      const emptyComponent = render({
+        projectRoot: "custom",
+        sources: I.Map()
+      }).component;
+      expect(emptyComponent).toMatchSnapshot();
+    });
+
+    it("calls clearProjectDirectoryRoot on click", async () => {
+      component.find(".sources-clear-root").simulate("click");
+      expect(props.clearProjectDirectoryRoot).toHaveBeenCalled();
+    });
+  });
+
+  describe("onContextMenu", () => {
+    let component;
+    let props;
+
+    const event = {
+      preventDefault: jest.fn(),
+      stopPropagation: jest.fn()
+    };
+
+    const directory = {
+      contents: new Map(),
+      name: "folder",
+      path: "folder/"
+    };
+
+    const file = {
+      contents: new Map(),
+      name: "file",
+      path: "folder/file.js"
+    };
+
+    const rootDirectory = {
+      contents: new Map(),
+      name: "root",
+      path: "root/"
+    };
+
+    beforeEach(() => {
+      const rendered = render({
```

Comment:
a better pattern for this would be to create an overrides object for this, rather than doing this in the before each. Component and props are returned by `render` by default. having this colocated in a test makes it easier to read than having it in a beforeEach